### PR TITLE
App insights

### PR DESCRIPTION
**Pull request by Smarker**
Added appinsights to my code from #6. Please let me know if there are any other areas where you think appinsights should be used or if you have suggestions for a different approach.

Added **APPINSIGHTS_INSTRUMENTATIONKEY** to env var.

### CODE REVIEW

File: src/storageClients/BlobStorageFactsManager.js
Comment (jcjimenez) [inline src/storageClients/BlobStorageFactsManager.js]: I know you didn't change this, but do you know if we want to do something other than throw on a potential invalid index from a malformed id?

**Review by jcjimenez — APPROVED** [review]
LGTM

**Comment by c-w** [comment]
Given that this is based on #6, I've temporarily changed the base of this review branch to #6's branch so that the changes in this pull request are made more obvious. There are a few ways forward from here:

1) After getting sufficient sign-offs, change the base to master again.
2) After getting a sign-off, merge the changes into #6.
3) Wait for #6 to be merged and then merge this.

File: src/connectors/CassandraConnector.js
Comment (c-w) [inline src/connectors/CassandraConnector.js]: I wonder if we can make this more generic/re-usable. Pseudo-code:

```js
withMetric('Cassandra connection time', () => new Promise((resolve, reject) => {
  client.connect()
  .catch((err) => {
    client.shutdown();
	reject(err);
  });
});

function withMetric(metricName, promiseAction) {
  let start = Date.now();
  promiseAction().then((val) => {
    let duration = Date.now() - start;
	iclient.trackMetric(metricName, duration);
	resolve(val);
  })
  .catch((err) => {
    iclient.trackException(err)
	reject(err);
  });
}
```

Adding a slight level of abstraction on top of using the naked client also has the advantage of separating the logging implementation from the code that does the business logic. Here, everything is still fine, but it's a slippery slope and easy to get to the situation where the logging/argument-checking code takes up the majority of a method and drowns out the business logic parts. Using this decorator style to handle logging will set a good example to prevent that.

Maybe it's worth spending 15 minutes to see if we can factor out the logging code like this, but it's not a high priority item so don't worry if you run into a roadblock.

File: src/resolvers/Topics.js
Comment (c-w) [inline src/resolvers/Topics.js]: We have this code repeated in quite a few places. Would it be possible to encapsulate this in one place to reduce the duplication (e.g. `src/loggingClient`)?

This has the added benefit that we can introduce a null-object ApplicationInsights client if the instrumentation key is not set so that we can enable running the app locally without requiring the instrumentation key to be specified which is a better developer experience.

Pseudo-code:

```js
const ikey = process.env.APPINSIGHTS_INSTRUMENTATIONKEY

let client;
if (ikey) {
  appInsights.setup(ikey);
  client = appInsights.getClient();
} else {
  client = {
    trackException: (ex) => console.log("exception " + ex),
    trackEvent: (route, details) => console.log("event " + route + " " + JSON.stringify(details))
  };
}

module.exports = {
  client: client
};
```

**Review by c-w — APPROVED** [review]

File: src/connectors/CassandraConnector.js
Comment (c-w) [inline src/connectors/CassandraConnector.js]: Please also document the key in the README.

Comment (erikschlegel) [reply src/connectors/CassandraConnector.js]: This is unnecessary as the  `applicationinsights` package automatically loads the instance based on the `APPINSIGHTS_INSTRUMENTATIONKEY` env variable https://github.com/Microsoft/ApplicationInsights-node.js#getting-started

File: src/connectors/CassandraConnector.js
Comment (erikschlegel) [inline src/connectors/CassandraConnector.js]: You should check that the `CASSANDRA_CONTACT_POINTS ` env variable is defined before it's used, and raise an app insight exception.

File: src/connectors/CassandraConnector.js
Comment (Smarker) [reply src/connectors/CassandraConnector.js]: ok!

File: src/resolvers/Topics.js
Comment (Smarker) [reply src/resolvers/Topics.js]: yes, good idea

File: src/connectors/CassandraConnector.js
Comment (erikschlegel) [inline src/connectors/CassandraConnector.js]: I believe these all are enabled by default

File: src/connectors/CassandraConnector.js
Comment (Smarker) [reply src/connectors/CassandraConnector.js]: good idea, I will try to do this

**Comment by Smarker** [comment]
I edited these files to address the comments for this PR:
* connectors/CassandraConnector -> executeQueries
* resolvers/Settings.js -> createOrReplaceSite (using insertSeedTopics)
* storageClients/BlobStorageManager -> getTopicsBySiteType
* storageClients/CassandraTableStorageManager -> prepareInsertTopic, batch (using chunkQueryPromises)

I addressed the comments by
* removing all instances of bluebird
* adding a connection pooling option to the Cassandra client
* making Cassandra client a singleton
* creating getTopicsBySiteType using a http request rather than BLOB_STORAGE_CONNECTION_STRING
* editing the insert topics statement
* adding app insights within the functions to track

File: src/connectors/CassandraConnector.js
Comment (erikschlegel) [inline src/connectors/CassandraConnector.js:17]: How did you come up with these settings?

File: src/connectors/CassandraConnector.js
Comment (erikschlegel) [inline src/connectors/CassandraConnector.js:3]: re: my earlier comment, setting `APPINSIGHTS_INSTRUMENTATIONKEY` is redundant as the `applicationinsights` library automatically does this for you

File: src/connectors/CassandraConnector.js
Comment (erikschlegel) [inline src/connectors/CassandraConnector.js:9]: Can you comment your code a bit more so other developers have more context into how this driver works(i.e. connection pooling, etc)

File: src/connectors/CassandraConnector.js
Comment (erikschlegel) [inline src/connectors/CassandraConnector.js:35]: I'm confused, why would one want to execute a batch of queries? Not sure if this is meant to address a batch of mutation statements(`insert`, `update`). If yes, please name your functions and variables in a way that helps other developers better understand your code.

File: src/connectors/CassandraConnector.js
Comment (erikschlegel) [inline src/connectors/CassandraConnector.js:41]: I think this will be tracked in app insights a key value bag of `properties` and not `measurements`. Is that your intention here?

File: src/connectors/CassandraConnector.js
Comment (erikschlegel) [inline src/connectors/CassandraConnector.js:42]: How does the connection get returned to the pool and/or closed?

File: src/connectors/CassandraConnector.js
Comment (erikschlegel) [inline src/connectors/CassandraConnector.js:26]: Can this throw an exception if none of the Contact points are available? Not sure if it does or not.

File: src/resolvers/Settings.js
Comment (erikschlegel) [inline src/resolvers/Settings.js:31]: Where is `Promise` imported?

File: src/resolvers/Settings.js
Comment (erikschlegel) [inline src/resolvers/Settings.js:30]: Lets create a new resolver called `CreateNewSite` and rename `createOrReplaceSite`->`ReplaceSite` and refactor accordingly. Both resolvers have two very different code paths and the code quality is starting to suffer because of it.

File: src/storageClients/BlobStorageManager.js
Comment (erikschlegel) [inline src/storageClients/BlobStorageManager.js:15]: the root context of the central host should be coming from the environment and not hard coded, as this is already available via the deployment.

File: src/resolvers/Settings.js
Comment (erikschlegel) [inline src/resolvers/Settings.js:308]: Im not sure I fully understand how this is working, as the json schema in the blob containers dont match up with how you're binding `topics` to the cassandra prepared statement ```{"id": 62, "topic": "climate", "value": "#climate"}, {"id": 63, "topic": "climate", "value": "climatechange"}, {"id": 64, "topic": "climate", "value": "globalwarming"}```

Comment (erikschlegel) [reply src/resolvers/Settings.js:308]: Also, I dont see any unit tests in your PR, which I would assume would catch this as from the surface this looks like it would fail.

File: src/resolvers/Settings.js
Comment (erikschlegel) [reply src/resolvers/Settings.js:30]: You'll also need to do some checks in the event the `CreateNewSite` is called on an already existing site, which in this case would continually add duplicate topics.

File: src/storageClients/BlobStorageManager.js
Comment (erikschlegel) [inline src/storageClients/BlobStorageManager.js:3]: Please remove `APPINSIGHTS_INSTRUMENTATIONKEY` reference. https://github.com/Microsoft/ApplicationInsights-node.js#getting-started

File: src/storageClients/BlobStorageManager.js
Comment (erikschlegel) [inline src/storageClients/BlobStorageManager.js:15]: you should be performing some validation on `siteType`

File: src/storageClients/BlobStorageManager.js
Comment (erikschlegel) [inline src/storageClients/BlobStorageManager.js:25]: the event name should be something thats short and consistent(`REQ_SITE`). You should be adding the url as a key entry in the `properties` map.

File: src/storageClients/BlobStorageManager.js
Comment (erikschlegel) [inline src/storageClients/BlobStorageManager.js:28]: I'd also log the ` response.statusCode` in your `errMsg`

File: src/storageClients/CassandraTableStorageManager.js
Comment (erikschlegel) [inline src/storageClients/CassandraTableStorageManager.js:6]: This insert statement doesnt belong in the table storage manager. Perhaps have a separate file to maintain your queries / mutation  statements.

File: src/storageClients/CassandraTableStorageManager.js
Comment (erikschlegel) [inline src/storageClients/CassandraTableStorageManager.js:5]: It's a bit difficult to follow the `chunkQueryPromises ` method. I mentioned using the `async` forEachLimit` in an earlier comment. Would probably help simplify things a bit.

File: src/storageClients/CassandraTableStorageManager.js
Comment (erikschlegel) [inline src/storageClients/CassandraTableStorageManager.js:34]: the function name of `batch` seems a bit vague.

File: src/storageClients/CassandraTableStorageManager.js
Comment (erikschlegel) [inline src/storageClients/CassandraTableStorageManager.js:8]: spark should be the only process writing to this table. I think you meant to write to `watchlist`

File: test/CassandraConnector.js
Comment (erikschlegel) [inline test/CassandraConnector.js:15]: there is no table called `Topics` https://github.com/CatalystCode/fortisdeploy/blob/119d3d847826b2cbd45791b3cf41c2ce1d60144d/ops/storage-ddls/cassandra-setup.cql

**Review by erikschlegel — CHANGES_REQUESTED** [review]
I'd like to better understand how the data quality for the results of these changes are being validated and unit tested. 

Lets sync up when you have a minute so we can walk through the comments. Probably makes sense to only focus on this feature until we're at a point where it's ready for a production release.

**Pull request closed without merging**
